# Huggingface에서 쉽게 트랜스포머 모델 써보기
안녕하세요! 강사 변형호입니다.   
이 실습에서는 지난 시간에 2주에 걸쳐 열심히 배웠던 트랜스포머 기반의 모델을 코드 한 줄로 사용해 보겠습니다.

**허깅페이스**[https://www.huggingface.co/] 는 트랜스포머와 같은 인공지능 모델을 쉽게 사용할 수 있는 서비스입니다.  
개발자가 자신의 모델을 자유롭게 업로드하고 공유할 수 있는 오픈 소스 플랫폼이고, 트랜스포머뿐만 아니라 최근에 사용되는 거의 모든 인공지능 모델들은 다 여기서 확인할 수 있다고 봐도 되는데요.     
이번에는 트랜스포머 모델을 가져와서 간단한 실습을 실행해 보겠습니다.

* 이 실습 파일은 https://huggingface.co/learn/nlp-course/ko/chapter1/3?fw=pt 를 참고하여 만들었습니다.




저희의 실습 환경에서는 따로 설치가 되어 있지 않기 때문에 아래의 코드를 실행하여 설치하겠습니다.

In [2]:
!pip install transformers

Defaulting to user installation because normal site-packages is not writeable
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.7/7.7 MB 12.1 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.0/302.0 kB 50.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 90.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 67.2 MB/s eta 0:00:00:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 166.4/166.4 kB 32.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 295.0/295.0 kB 54.4 MB/s eta 0:00:00

[notice] A new release of pip available: 22.3.1 -> 23.3.1
[notice] To update, run: pip install --upgrade pip


설치 후에는 import를 통해 불러옵니다.  
pipeline은 중간 과정을 생략하고 이미 만들어진 모델을 통해 원하는 문제를 빠르게 실행해 볼 수 있습니다.

In [6]:
from transformers import pipeline

파이프라인에 문제를 설정하고 문장을 입력해 주면 기본 모델에 의한 결과를 돌려줍니다.

아래의 문제들을 실행해 보겠습니다.   

sentiment-analysis : 감정 분석   
text-generation : 텍스트 생성   
fill-mask : 마스크 채우기   
question-answering : 질의 응답   
summarization : 요약   

zero-shot-classification : 제로샷 분류(규칙을 스스로 찾아서 학습하는 문제)    

In [17]:
# 감정 분석 : 문장을 입력받으면 긍정/부정을 분류하여 출력
classifier = pipeline("sentiment-analysis") 
classifier("I had a really delightful meal.")

No model was supplied, defaulted to distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
All PyTorch model weights were used when initializing TFDistilBertForSequenceClassification.

All the weights of TFDistilBertForSequenceClassification were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFDistilBertForSequenceClassification for predictions without further training.


[{'label': 'POSITIVE', 'score': 0.9998764991760254}]

I had a really delightful meal. 이라는 문장은 99.99%로 긍정적인 문장이라고 분류했습니다.<br><br>

In [19]:
# 텍스트 생성 : 문장을 입력받아 뒷부분을 추론하여 출력 (성능은 별로...)
generator = pipeline('text-generation')
generator('My professor does not like my')

No model was supplied, defaulted to gpt2 and revision 6c0e608 (https://huggingface.co/gpt2).
Using a pipeline without specifying a model name and revision in production is not recommended.
All PyTorch model weights were used when initializing TFGPT2LMHeadModel.

All the weights of TFGPT2LMHeadModel were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFGPT2LMHeadModel for predictions without further training.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'My professor does not like my grades. I don\'t really understand how my grades make sense."\n\nHe added: "But there are certain things you can do to earn a good GPA – like writing a good job, doing things that help people'}]

문장의 뒷 내용을 추론하여 채워 줍니다.   
텍스트 생성의 기본 모델은 gpt2입니다. (붉은 warning message에도 나와 있습니다.) <br><br><br>

In [21]:
# 마스크 채우기 : 빈칸에 무슨 단어가 들어갈 지 추론하기
unmasker = pipeline('fill-mask')
unmasker('We have to <mask> balanced meal.')

No model was supplied, defaulted to distilroberta-base and revision ec58a5b (https://huggingface.co/distilroberta-base).
Using a pipeline without specifying a model name and revision in production is not recommended.
All PyTorch model weights were used when initializing TFRobertaForMaskedLM.

All the weights of TFRobertaForMaskedLM were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFRobertaForMaskedLM for predictions without further training.


[{'score': 0.6801381707191467,
  'token': 3529,
  'token_str': ' eat',
  'sequence': 'We have to eat balanced meal'},
 {'score': 0.030325133353471756,
  'token': 2254,
  'token_str': ' enjoy',
  'sequence': 'We have to enjoy balanced meal'},
 {'score': 0.023026229813694954,
  'token': 2807,
  'token_str': ' choose',
  'sequence': 'We have to choose balanced meal'},
 {'score': 0.021803976967930794,
  'token': 1306,
  'token_str': ' ensure',
  'sequence': 'We have to ensure balanced meal'},
 {'score': 0.02157534658908844,
  'token': 146,
  'token_str': ' make',
  'sequence': 'We have to make balanced meal'}]

mask로 표시된 부분에 맞는 후보 단어를 찾아 점수 순으로 출력합니다.  <br><br><br>

In [26]:
# 질의응답 : 질문에 맞는 답변하기 (성능은 별로...)
# Context와 Question을 넣어서 상황과 질문을 전달

qa = pipeline('question-answering')
qa(question='Which class did I fail?',
  context='I got an F in math class because I have missed too many classes')

No model was supplied, defaulted to distilbert-base-cased-distilled-squad and revision 626af31 (https://huggingface.co/distilbert-base-cased-distilled-squad).
Using a pipeline without specifying a model name and revision in production is not recommended.
All PyTorch model weights were used when initializing TFDistilBertForQuestionAnswering.

All the weights of TFDistilBertForQuestionAnswering were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFDistilBertForQuestionAnswering for predictions without further training.


{'score': 0.9106036424636841, 'start': 14, 'end': 18, 'answer': 'math'}

상황과 질문을 주고 간단한 질의응답을 수행합니다.    
수학 수업을 너무 많이 빼먹어서 F를 받았다는 상황에서, 정확한 답변을 하고 있습니다.<br><br><br>

In [28]:
# 요약 : 긴 글 요약하기

summarizer = pipeline("summarization")  # 요약
summarizer("""
Transformers is a 2007 American science fiction action film based on Hasbro's toy line of the same name. 
It is the first installment in the Transformers film series. 
The film is directed by Michael Bay from a screenplay by Roberto Orci and Alex Kurtzman. 
It stars Shia LaBeouf as Sam Witwicky, a teenager who gets caught up in a war between the heroic Autobots and the villainous Decepticons, two factions of alien robots who can disguise themselves by transforming themselves into everyday machinery, primarily vehicles. 
The Autobots intend to retrieve and use the AllSpark, the powerful artifact that created their robotic race that is on Earth, to rebuild their home planet Cybertron and end the war, while the Decepticons have the intention of using it to build an army by giving life to the machines of Earth. 
Tyrese Gibson, Josh Duhamel, Anthony Anderson, Megan Fox, Rachael Taylor, John Turturro and Jon Voight also star, while Peter Cullen and Hugo Weaving voice Optimus Prime and Megatron, respectively.
""")

No model was supplied, defaulted to t5-small and revision d769bba (https://huggingface.co/t5-small).
Using a pipeline without specifying a model name and revision in production is not recommended.


All PyTorch model weights were used when initializing TFT5ForConditionalGeneration.

All the weights of TFT5ForConditionalGeneration were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFT5ForConditionalGeneration for predictions without further training.


2023-10-29 06:55:45.906442: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x55bf888a45e0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2023-10-29 06:55:45.908687: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2023-10-29 06:55:46.277462: I tensorflow/compiler/mlir/tensorflow/utils/dump_mlir_util.cc:255] disabling MLIR crash reproducer, set env var `MLIR_CRASH_REPRODUCER_DIRECTORY` to enable.
2023-10-29 06:55:47.194941: I ./tensorflow/compiler/jit/device_compiler.h:186] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.


[{'summary_text': "Transformers is a 2007 american science fiction action film based on Hasbro's toy line of the same name . it stars Shia LaBeouf as Sam Witwicky, a teenager who gets caught up in a war between the heroic Autobots and the villainous Decepticons . the autobots intend to retrieve and use the AllSpark to rebuild their home planet Cybertron and end the war ."}]

영화 트랜스포머에 대한 한 문단의 글을 잘 요약했습니다. <br><br><br>

In [30]:
# 번역 (한국어 --> 영어)
translator2 = pipeline("translation", model="Helsinki-NLP/opus-mt-ko-en")  
translator2("트랜스포머는 현대 AI 모델의 기반이 되었습니다.")
 

All model checkpoint layers were used when initializing TFMarianMTModel.

All the layers of TFMarianMTModel were initialized from the model checkpoint at Helsinki-NLP/opus-mt-ko-en.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFMarianMTModel for predictions without further training.


[{'translation_text': 'Transpomer became the basis for the modern AI model.'}]

번역의 경우 한국어 문장도 학습이 되어 있어야 하므로 다른 모델[https://huggingface.co/Helsinki-NLP/opus-mt-ko-en] 을 명시하여 불러왔습니다.   
transformers의 pipeline을 이용하면 이렇게 특정 모델을 다운로드하여 바로 활용할 수 있습니다. <br><br><br>

In [40]:
# 제로 샷 분류 : 아무런 사전정보 없이 분류시키기

zero_shot_classifier = pipeline("zero-shot-classification")
zero_shot_classifier(
    "I eat banana!",
    candidate_labels=["Elemantary level", "Middle school level", "High school level", 'College/university level'],
)

No model was supplied, defaulted to roberta-large-mnli and revision 130fb28 (https://huggingface.co/roberta-large-mnli).
Using a pipeline without specifying a model name and revision in production is not recommended.
All PyTorch model weights were used when initializing TFRobertaForSequenceClassification.

All the weights of TFRobertaForSequenceClassification were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFRobertaForSequenceClassification for predictions without further training.


{'sequence': 'I eat banana!',
 'labels': ['Elemantary level',
  'Middle school level',
  'High school level',
  'College/university level'],
 'scores': [0.34958964586257935,
  0.22910840809345245,
  0.22269058227539062,
  0.198611319065094]}

또한, zero-shot-classification (제로샷 분류)를 수행할 수 있는데요.   
이것은 사전에 저장된 방법이 아니라, 단순히 문장과 분류 후보를 주면    
트랜스포머가 추론을 통해 답을 찾는 방식입니다.   
기본값으로 큰 모델을 불러오기 때문에 시간이 조금 소요됩니다.    

제가 낸 문제는 '이 말을 하는 사람이 초등/중/고등/대 학생 중에 어떤 사람일까?' 를 맞추는 문제였는데요.    
예제에서는 대략 맞췄지만 정확히 이해하지는 못한 것 같습니다.

이렇게 기존 모델을 불러와서 사용하는 것도 물론 가능하지만,    
실제 환경에서는 사용자가 쓰고 싶은 데이터에 맞게 트랜스포머 모델의 가중치를 변화시키는 방법을 사용합니다.    
이를 파인 튜닝(Fine-tuning)이라고 하는데요,    
저희가 이전에 fasttext로 ko100.bin을 학습시켜서 얻은 임베딩에 건축 관련 데이터를 추가했던 것과 유사합니다.   

단, 파인 튜닝은 모델이 클수록 시간이 오래 걸리므로 최신의 GPT-3.5와 같은 거대 언어 모델에서는 다른 방법들도 함께 사용합니다.

다음 실습 시간에는 트랜스포머와 GPT를 사용하여 실제 챗봇을 구현해 보겠습니다.